program/src/processor.rs

```diff
@@ -5313,6 +5313,91 @@ impl Processor {
         Ok(())
     }
 
+    fn flash_loan(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        liquidity_amount: u64,
+    ) -> MangoResult {
+        const NUM_FIXED: usize = 6;
+        let (fixed_ais, remaining_ais) = array_refs![accounts, NUM_FIXED; ..;];
+        let [
+            mango_group_ai,           // read
+            signer_ai,                // read
+            source_token_ai,          // write
+            destination_token_ai,     // write
+            token_prog_ai,            // read
+            flash_loan_prog_ai,       // read
+        ] = fixed_ais;
+
+        check!(program_id != flash_loan_prog_ai.key, MangoErrorCode::InvalidProgramId)?;
+
+        let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        let source_token = Account::unpack(&source_token_ai.try_borrow_data()?)?;
+
+        let flash_loan_amount = if liquidity_amount == u64::MAX {
+            source_token.amount
+        } else {
+            min(source_token.amount, liquidity_amount)
+        };
+
+        let source_balance_before = source_token.amount;
+
+        let mut flash_loan_instruction_accounts = vec![
+            AccountMeta::new(*destination_token_ai.key, false),
+            AccountMeta::new(*source_token_ai.key, false),
+            AccountMeta::new_readonly(*token_prog_ai.key, false),
+        ];
+        let mut flash_loan_instruction_account_infos = vec![
+            destination_token_ai.clone(),
+            source_token_ai.clone(),
+            token_prog_ai.clone(),
+            flash_loan_prog_ai.clone(),
+        ];
+        for account_info in remaining_ais.iter() {
+            flash_loan_instruction_accounts.push(AccountMeta {
+                pubkey: *account_info.key,
+                is_signer: account_info.is_signer,
+                is_writable: account_info.is_writable,
+            });
+            flash_loan_instruction_account_infos.push(account_info.clone());
+        }
+
+        let signers_seeds = gen_signer_seeds(&mango_group.signer_nonce, mango_group_ai.key);
+        invoke_transfer(
+            token_prog_ai,
+            source_token_ai,
+            destination_token_ai,
+            signer_ai,
+            &[&signers_seeds],
```

Comment:
If we sign like this and allow the program arbitrary writable accounts via `remaining_ais`, then what keeps the program from affecting other mango accounts guarded by this signature?

Reply:
Nevermind, since this is just the transfer and not the cpi call it's fine.